### ISSUE TITLE
[BLOCKED by #128 + #129] [Audit Correction 3/9] Approval freshness + policy/preflight + retry handoff closure

### ISSUE THREAD

**Issue by bonggyulim** [solar-ai-dev/mcp-work-agent#130]
# Audit Correction 3/9 — Approval freshness + policy/preflight + retry handoff closure

Blocked by: `#128`, `#129`.

## Scope

Close the Approval/policy/preflight/retry-handoff correction boundary represented by:

```text
002-03  ExpireApproval lacks exact TTL/snapshot-stale eligibility
025-01  WritePreflight bypasses ExpireApproval → RefreshExpiredAction lifecycle
019-09  EvaluateActionPolicy over-blocks evidence-backed modifications
D-011   PrepareWriteRetry omits required REVIEW_ENTRY durable handoff
```

Treat these as one capability chain. Approval freshness, policy decision authority, preflight lifecycle, Review state, WorkflowHandoff and production callers must all implement the same Canonical semantics.

## Authority and implementation derivation

Use this order for every decision:

```text
Canonical concern owners
→ implementation-inventory/ledger.md
→ implementation-inventory/canonical-current-implementation-map.md
→ actual current source / callers / imports / exports / repositories / tests
```

Applicable Canonical reading set:

```text
00-project-source-guide.md
01-b-policy-definition.md
04-a-domain-state-transition-contract.md
04-domain-database-design.md
07-tool-mcp-internal-interface.md
08-sequence-design.md
11-observability-logging-audit.md
12-a-state-transition-test-matrix.md
12-test-design.md
16-repository-architecture/**
```

Interpret the layers as follows:

- **Canonical** defines Approval lifecycle, policy alternatives, freshness semantics, handoff requirements and semantic ownership.
- **Ledger** identifies the required implementation obligations/capabilities.
- **Mapping is the most concrete implementation authority** for the repository: exact final path/symbol, policy/Approval owner, repository/handoff owner, reuse/move/merge/split disposition, production callers and test owner.
- If Ledger is incomplete or ambiguous and Mapping resolves it consistently with Canonical, follow the Mapping resolution.
- If Mapping conflicts with Canonical, Canonical wins and Mapping must be corrected.
- Actual source/callers prove whether the Mapping is true in production.

Before editing each concept, resolve it through Canonical → Ledger → Mapping and census current definitions plus all production callers for ExpireApproval, RefreshExpiredAction, WritePreflight, EvaluateActionPolicy, PrepareWriteRetry, Review/Approval repositories, confirmation evidence and WorkflowHandoff. Reuse the exact mapped/current authority when it already exists. Introduce a new DTO/Enum/Literal/policy helper/repository/handoff owner only when Canonical requires a distinct concept, Ledger includes the obligation, Mapping places it as a distinct production authority, and repository-wide census proves no reusable current implementation exists.

### Approval freshness

Implement `ExpireApproval` using the exact current eligibility derived through Canonical → Ledger → Mapping: current Approval/Plan/Action authority plus TTL and relevant source/policy/schema/binding staleness. A still-valid ACTIVE Approval cannot be expired arbitrarily. Replay/request-hash semantics must use the mapped canonical command authority.

### WritePreflight lifecycle

When fresh preflight evidence invalidates the approved snapshot, use the mapped canonical lifecycle:

```text
ExpireApproval
→ RefreshExpiredAction
→ fresh review / reapproval
```

Reuse already-fetched provider evidence when valid. Child mutation, Receipt/Audit and required continuation semantics must remain transactionally correct. The preflight path cannot retain a second direct Modify/revoke lifecycle authority.

### Policy decision authority

Implement `EvaluateActionPolicy` as the single mapped deterministic policy authority for current evidence and policy facts. Support all Canonical evidence-backed modification and confirmation alternatives. Confirmation evidence must remain bound to the current Action/version/arguments/source/evidence/policy context. Agent/LLM output cannot become policy authority.

### PrepareWriteRetry continuation

An applied retry preparation must stage the mapped canonical `REVIEW_ENTRY` continuation in the same short UoW as the owning Action/Review mutation, Receipt and required Audit. Workflow executor scheduling occurs only after commit. Replay must not duplicate handoff or required Audit facts.

## Chain-wide impact

Inspect and correct:

```text
Domain Action/Approval guards
ExpireApproval / RefreshExpiredAction
WritePreflight and claim-stale recovery callers
EvaluateActionPolicy and confirmation evidence
PrepareWriteRetry
Approval / Review repositories
WorkflowHandoffRepository
UnitOfWork participant ordering
API/Application retry callers
replay/restart behavior
all imports/exports of corrected policy/Approval semantic types
```

Use the Mapping to locate exact current/final paths and symbols, then verify those Mapping claims against actual production callers before editing.

Preserve existing correct implementation wherever possible:

```text
KEEP → MOVE/RENAME → SPLIT/MERGE → TARGETED_CORRECTION
→ REWRITE only when reuse is unsafe
→ DELETE only after caller cut-over
```

## Validation

Run validation proportional to the actual change set:

1. focused pytest covering all owned Approval/policy/preflight/handoff defects and negative cases;
2. directly related regression tests for changed repositories, preflight and callers;
3. changed-scope Ruff on modified Python files;
4. changed-scope mypy on modified typed production modules where supported;
5. replay/rollback/handoff tests required by the changed transaction behavior;
6. repository census proving intended callers use the mapped canonical authority and old caller/import/export/duplicate writer counts are zero;
7. final re-read of changed Mapping rows against the actual pushed production SHA.

Repository-wide pytest/Ruff/mypy is used only when a changed shared primitive has a blast radius that focused and directly related suites cannot establish.

### Validation cycle

After implementation, perform one complete Issue-scope validation covering all owned defects, the full corrected chain, focused tests and negative proof. If that round discovers real defects, fix all defects found in that round and perform one additional Issue-scope validation. Multiple defects found in one round still add only one validation round. Finish after one clean Issue-scope validation round.

## Completion gate

```text
all owned Approval/policy defects closed
Canonical → Ledger → Mapping chain reconciled
Mapping final path/symbol/caller disposition matches actual production state
Approval expiry authority exact
preflight lifecycle bypass = 0
policy decision authority = 1
confirmation context mismatch acceptance = 0
retry continuation loss = 0
old caller/import/export/duplicate authority = 0
focused pytest = PASS
changed-scope Ruff = PASS
changed-scope mypy = PASS where applicable
required replay/rollback/handoff checks = PASS
```

After the production correction is pushed, update `implementation-inventory/canonical-current-implementation-map.md` to the actual final production SHA and exact final path/symbol/caller state. Change `ledger.md` only if the Canonical implementation universe itself changed.

**Comment by bonggyulim** [comment]
#130 완료 보고

- execution start SHA: `a2e42eb116ef8703aa527afe1347000e57353c6b`
- final production SHA: `e91b123d195aaa534699ea42cf49984a0659071f`
- final Map SHA: `b56916fa82c5329fbd252138de381a73445ff8c5`
- local HEAD == remote HEAD, worktree clean

## Defect closure

- `002-03`: caller hash만 받던 ExpireApproval을 current source snapshot 입력 + handler canonical hash 재계산으로 좁혔다. Domain guard가 current Plan/Action/Approval authority와 TTL, action version, arguments, source, policy, schema stale set을 exact 검사하며 still-current ACTIVE Approval은 거부한다.
- `025-01`: WritePreflight의 direct ModifyAction + approval revoke 경로를 제거했다. fresh provider evidence를 재사용해 `ExpireApproval -> RefreshExpiredAction -> Review REQUIRED + REVIEW_ENTRY handoff`로 전환했고, current deterministic feasibility DENY는 refresh가 아니라 `BlockRun`으로 보낸다.
- `019-09`: EvaluateActionPolicy를 ALLOW/DENY/CONFIRMATION_REQUIRED 단일 결정 authority로 정리했다. user-selected target, independent Evidence 2개, explicit Gmail/Resource relation 대안을 지원하고 confirmation hash를 Action/version/arguments/source/Evidence/policy에 묶었다. stale/unreferenced/mismatched receipt는 허용하지 않는다.
- `D-011`: PrepareWriteRetry가 Action/Review/Receipt/Audit와 PENDING `MAIN_CONTROL:REVIEW_ENTRY` WorkflowHandoff를 같은 UoW에서 commit한다. executor schedule은 commit 이후에만 호출되고 replay는 handoff/audit를 중복 생성하지 않는다.

## Chain corrections

- stale refresh가 fresh source/policy/schema/risk와 target ResourceRef version을 함께 재결합한다.
- preflight와 claim-stale recovery가 같은 Expire/Refresh handler instance를 사용한다.
- planning/save/modify/preflight policy caller가 shared evidence decision kernel을 사용한다.
- API retry caller가 UUID/resume registry/scheduler canonical dependencies를 주입한다.
- applied migration 변경 0, `ledger.md` 변경 0. 새 migration도 필요하지 않았다.

## Negative proof

- valid non-stale arbitrary expiry = 0
- preflight direct Modify/revoke bypass = 0
- false target-only policy block = 0
- confirmation context mismatch accepted = 0
- PrepareWriteRetry applied without REVIEW_ENTRY handoff = 0
- handoff stage failure partial commit = 0
- same-command replay duplicate handoff/audit = 0
- 각 Issue-owned production authority definition = 1
- old caller/import/export/duplicate authority = 0

## Validation

- focused Issue suites: `212 passed`
- final full suite: `2478 passed, 6 skipped`
- changed-file Ruff: PASS
- changed production mypy (기존 baseline 오류 파일 제외): 12 files PASS
- compileall: PASS
- `git diff --check`: PASS
- Map integrity: CAP 142 + STR 473 + NPA 85 = 700; missing 0, extra 0, duplicate 0; Map ID set == Ledger ID set

Complete validation passes: 3

1. pass 1: confirmed four owned defects plus fresh ResourceRef rebinding and deterministic DENY -> BlockRun linkage.
2. pass 2: found and fixed independent-Evidence provenance, explicit Gmail/Resource relation propagation, raw source-hash expiry input, and confirmation Evidence hash binding.
3. pass 3: new root defect 0; all completion gates clean.

External-owner remainder: #130 scope에는 없음. 저장소 전체 mypy에는 이번 변경 hunk 밖의 기존 9건(`execution_phase.py`, `workflow.py`)이 남아 있으나 #130 focused changed-source type check와 전체 runtime test는 통과했다.

**Comment by bonggyulim** [comment]
## Canonical → Ledger → Current Map adherence audit finding

#130은 기존 behavior-focused 완료 판정과 별개로, 아래 hierarchy로 다시 보면 **현재 그대로 `VERIFIED_COMPLETE`로 사용할 수 없습니다.**

```text
Canonical semantic contract
→ ledger.md exact implementation universe / symbol target
→ canonical-current-implementation-map.md current realization
→ actual production code
```

### Authority chain

`00 Project Source Guide`는 Policy semantics를 `01-B`, Workflow artifact ownership을 `06`, typed interface를 `07`, repository path/file/symbol realization을 `16`에 둡니다. `16/01 Spec → Code Deterministic Mapping`은 implementation task가 semantic specification에서 시작해야 하며, second live semantic authority가 생기면 `SEMANTIC_AUTHORITY_COLLISION`으로 중단하도록 합니다.

Ledger의 `CAP-APP-037`은 `action.evaluate_action_policy`의 canonical Application symbols를 정확히:

```text
EvaluateActionPolicyQueryV1
ActionPolicyEvaluationResultV1
EvaluateActionPolicyHandler
```

로 닫고, behavior source를 `01-B Policy Definition`으로 지정합니다.

Current Map의 `CAP-APP-037`은 더 구체적으로:

- `EvaluateActionPolicyHandler`가 sole deterministic policy decision matrix
- confirmation context가 current Action/Evidence/source/policy에 exact binding
- duplicate policy kernel 제거

를 current production realization으로 선언합니다.

### Concrete mismatch — competing/surrogate confirmation contract

그런데 current production `application/use_cases/action/evaluate_action_policy.py`는 추가 public semantic artifacts를 정의/export합니다.

```python
PolicyConfirmationKind = Literal[
    "SCOPE_EXPANSION",
    "DUPLICATE_OVERRIDE",
    "CONFLICT_OVERRIDE",
]

@dataclass(frozen=True, slots=True)
class PolicyConfirmationEvidenceV1:
    receipt_ref: str
    confirmation_kind: PolicyConfirmationKind
    decision: Literal["APPROVED", "DECLINED"]
    decision_context_hash: str
```

그리고 `PolicyConfirmationEvidenceV1` 및 confirmation-hash helper를 module `__all__`로 public export합니다.

하지만 Canonical 06은 이미 동일 policy-confirmation fact의 canonical artifact를 **`PolicyConfirmationReceiptV1`** 하나로 정의하며, 이 타입에는 최소 다음 authority/provenance가 포함됩니다.

```text
schema_version
meta: StateArtifactMetaV1
interrupt_id
confirmation_kind
decision
semantic_owner_id
decision_context_hash
affected_route_ids
affected_resource_refs
```

Canonical 06은 이 Receipt가 실제 user interrupt response를 받은 Application/Confirmation Controller에서만 생성되고, `meta.based_on` freshness/provenance 및 semantic owner/affected refs를 보존해야 한다고 명시합니다.

따라서 현재 `PolicyConfirmationEvidenceV1`은 단순 private helper 구조가 아니라:

- canonical `PolicyConfirmationReceiptV1`과 같은 semantic fact를 별도 public V1 type으로 다시 표현하고,
- canonical receipt의 provenance/owner fields를 축약하고,
- Ledger `CAP-APP-037` exact symbol set에도 포함되지 않으며,
- Map은 오히려 `exact confirmation context` / `duplicate policy kernel removed`를 선언하고 있다는 점에서

**Canonical → Ledger → Map → code chain과 충돌하는 competing/surrogate contract 후보가 아니라 concrete architecture finding으로 봐야 합니다.**

### Classification

```text
Finding: F-130-CANONICAL-ADHERENCE-01
Classification: COMPETING_TYPE / LEDGER_MAP_DEVIATION
Affected defect: 019-09
Severity: architecture blocker until resolved
```

이 finding의 핵심은 `PolicyConfirmationEvidenceV1`이 유틸성이 낮거나 dataclass라서 문제가 아니라, **이미 canonical owner/type이 있는 semantic contract를 별도 public type으로 재설계했다는 것**입니다.

### Required correction direction

둘 중 하나만 가능합니다.

1. 현재 Canonical이 맞다면:
   - `PolicyConfirmationReceiptV1` canonical authority를 그대로 소비하고,
   - `PolicyConfirmationEvidenceV1` / duplicate public vocabulary를 제거하며,
   - evaluator에는 필요한 read-only projection만 private/internal detail로 남긴다.

2. 현재 code shape가 정말 새 canonical design이라면:
   - code를 근거로 합리화하지 말고,
   - 먼저 applicable Canonical owner를 변경하고,
   - Ledger를 재추출/revalidate하고,
   - Current Map을 새 target으로 갱신한 후 implementation을 맞춘다.

현재처럼 code만 다른 public contract를 갖고 Map이 `FULL/COMPLETE`를 선언하는 상태는 허용하면 안 됩니다.

### Verdict for this audit axis

`#130 = NOT_VERIFIED_COMPLETE_ON_CANONICAL_LEDGER_MAP_ADHERENCE`

기존 `002-03 / 025-01 / D-011` behavior correction이 잘못되었다는 의미는 아닙니다. 이 코멘트는 **019-09 policy contract realization이 Canonical → Ledger → Map chain을 exact하게 따르지 않았다는 별도 architecture blocker**를 기록합니다.

Issue state는 이 코멘트에서 변경하지 않습니다. 다만 Architecture final closure에서는 이 finding이 해소되기 전 #130을 최종 완료 근거로 사용하면 안 됩니다.
